### PR TITLE
Fixes for openssh

### PR DESCRIPTION
**Pull request by julek-wolfssl**

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -34814,14 +34817,33 @@ int wolfSSL_ECDH_compute_key(void *out, size_t outlen,
     }
 
     len = (word32)outlen;
+    key = (ecc_key*)ecdh->internal;
+
+#if defined(ECC_TIMING_RESISTANT) && !defined(HAVE_SELFTEST) \
+    && !defined(HAVE_FIPS)
+    if (key->rng == NULL) {
```
Comment (toddouska) [inline src/ssl.c:34824]: This does not look correct. What about the case case with key->rng != NULL, where is key->rng set then?

File: src/ssl.c [hunk 2]
```diff
@@ -34814,14 +34817,33 @@ int wolfSSL_ECDH_compute_key(void *out, size_t outlen,
...
+        if (initGlobalRNG == 0 && wolfSSL_RAND_Init() != WOLFSSL_SUCCESS) {
+            WOLFSSL_MSG("No RNG to use");
+            return WOLFSSL_FATAL_ERROR;
+        }
+        key->rng = &globalRNG;
```
Comment (toddouska) [inline src/ssl.c:34829]: Why the global?  Won't wolfSSL_EVP_PKEY_free() or some other free end up free'ing the globalRNG if something goes wrong before you "unset" it?

File: wolfcrypt/src/asn.c [hunk 3]
```diff
@@ -16286,6 +16288,20 @@ int wc_EccPrivateKeyToPKCS8(ecc_key* key, byte* output, word32* outLen)
...
+
+/* Write only private ecc key to unencrypted PKCS#8 format.
+ *
+ * return length on success else < 0 */
+int wc_EccPrivateKeyToPKCS8(ecc_key* key, byte* output, word32* outLen)
+{
+    return eccToPKCS8(key, output, outLen, 0);
+}
+
+int wc_EccKeyToPKCS8(ecc_key* key, byte* output,
```
Comment (toddouska) [inline wolfcrypt/src/asn.c:16300]: Please document, at least like the function above.

File: wolfcrypt/src/ecc.c [hunk 4]
```diff
@@ -4016,20 +4029,50 @@ int wc_ecc_shared_secret_ex(ecc_key* private_key, ecc_point* point,
...
+        WOLFSSL_MSG("wc_ecc_is_valid_idx failed");
         return ECC_BAD_ARG_E;
+    }
 
     switch(private_key->state) {
         case ECC_STATE_NONE:
         case ECC_STATE_SHARED_SEC_GEN:
             private_key->state = ECC_STATE_SHARED_SEC_GEN;
 
+            #ifdef ECC_TIMING_RESISTANT
```
Comment (toddouska) [inline wolfcrypt/src/ecc.c:4047]: Why is this needed now for shared_secret_gen?

File: src/ssl.c (hunk 1 above)
Comment (julek-wolfssl) [reply src/ssl.c:34824]: `key->rng` should be set by wc_ecc_set_rng if using ECC_TIMING_RESISTANT. Would you like me to move this to an initialization of key (ex: wolfSSL_EC_KEY_new)?

File: src/ssl.c (hunk 2 above)
Comment (julek-wolfssl) [reply src/ssl.c:34829]: I'll modify unsetting so that it always occurs after the wc_ecc_shared_secret_ssh call.

File: wolfcrypt/src/ecc.c (hunk 4 above)
Comment (julek-wolfssl) [reply wolfcrypt/src/ecc.c:4047]: This should be done by the user in `wc_ecc_set_rng` (or in the compatibility layer). I'll remove my changes to `wc_ecc_shared_secret_ex`.

File: wolfcrypt/src/evp.c [hunk 5]
```diff
@@ -6041,20 +6041,18 @@ int wolfSSL_EVP_PKEY_assign(WOLFSSL_EVP_PKEY *pkey, int type, void *key)
...
 static void ECC_populate_EVP_PKEY(EVP_PKEY* pkey, ecc_key* ecc)
 {
-    int ret;
+    word32 derSz = 0;
     if (!pkey || !ecc)
         return;
-    if ((ret = wc_EccPublicKeyDerSize(ecc, 1)) > 0) {
-        int derSz = ret;
-        char* derBuf = (char*)XMALLOC(derSz, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+    if (wc_EccKeyToPKCS8(ecc, NULL, &derSz) == LENGTH_ONLY_E) {
```
Comment (SparkiDev) [inline wolfcrypt/src/evp.c:6047]: MEMORY_E and such are being lost.
Need to make this function return an int.

**Review by SparkiDev — CHANGES_REQUESTED** [review]

**Comment by julek-wolfssl** [comment]
Failing valgrind test is probably solved in #3534 . Need to rebase once that is merged to get valgrind tests passing.

**Comment by douzzer** [comment]
The valgrind failure was indeed a false positive, but I don't think it's anything to do with #3534 or wolfSSL/testing#173.  I've seen this recurrent glitch on other PRs and retry cleared it -- valgrind missing `IntelRDrand64` instruction.  This time it's on the "master" node (test.wolfssl.com itself, Xeon E3-1225 v3), and I think that's where I've seen it before.  Retry is happening on master and it's just failing again.  Maybe need to mask out master for the test?

**Comment by julek-wolfssl** [comment]
Kaleb and I took a look into this issue and it looks like #3534 is not the solution. It looks like this test is failing on the master node always. My theory is that the valgrind version (3.11) is to blame (the version on the master node). I'm not sure what the valgrind version is on the other Jenkins nodes. The test passes on my machine with valgrind-3.15.0. Kaleb has passed this onto Chris for him to look at and see where the issue may be.

**Comment by douzzer** [comment]
I checked the valgrind versions on various Bozeman cluster hosts earlier, and they're mostly a mix of 3.10 and 3.11, so that doesn't seem to explain it.  That said, upgrading valgrind is an easy thing to try.

**Comment by kaleb-himes** [comment]
It may be that the chipset on master doesn’t have that instruction and it is bad to assume that --enable-intelrand is supported anywhere? That config option was only added to testing on Aug 27th so this has not been failing for long but it has been failing 100% on select slaves since the merge of this testing PR: https://github.com/wolfSSL/testing/commit/00fd5fe.

**Comment by kaleb-himes** [comment]
1 of 2 options?

- Remove the assumption `--enable-intelrand` will be supported on all slaves (Don't include it in the test config set for valgrind which can run on any slave)
- Restrict valgrind test to a finite subset of slaves and determine which ones have the instructions needed to support the configure option `--enable-intelrand`

**Comment by douzzer** [comment]
The Xeon in Bozeman master has `rdrand` in its /proc/cpuinfo flags, so I'd like to try a bit harder to make it work on master before we start kludge-masking it out.

Should we first try `apt-get update` and `apt-get install valgrind` on master, to see if that just fixes it there?


Btw, @julek-wolfssl , you wouldn't need to rebase your PR for it to pick up fixes in master for a Jenkins retry.  Jenkins always tests PRs with the current master as the base (applies the commits in the PR to current master), regardless of the current base in the branch-of-origin for the PR.  That's also why Jenkins can't test PRs with conflicts with current master.

**Comment by julek-wolfssl** [comment]
Yeah I forgot that Jenkins merges into master for testing.

A missing instruction on a specific chipset should cause a fault or exception on all tests running that assembly instruction (unless only valgrind tests intelrand) and not just valgrind.

**Comment by kaleb-himes** [comment]
retest this please. (FIPSv2 tests got cleared from logs and needs to re-run)

**Review by SparkiDev — APPROVED** [review]

**Comment by julek-wolfssl** [comment]
Retest this please.

**Review by toddouska — APPROVED** [review]
